### PR TITLE
Add Mut::clone_from_if_neq

### PR DESCRIPTION
**Pull request by SOF3**
# Objective

- Support more ergonomic conditional updates for types that can be modified by `clone_into`.

## Solution

- Use `ToOwned::clone_into` to copy a reference provided by the caller in `Mut::clone_from_if_neq`.

## Testing

- See doc tests.

### CODE REVIEW

File: crates/bevy_ecs/src/change_detection.rs [hunk 1]
```diff
@@ -268,6 +268,53 @@ pub trait DetectChangesMut: DetectChanges {
...
+    /// # schedule.add_systems(update_message);
+    /// #
+    /// # // first time `reset_score` runs, the score is changed.
+    /// # schedule.run(&mut world);
+    /// # assert!(message_changed.run((), &mut world));
+    /// # // second time `reset_score` runs, the score is not changed.
+    /// # schedule.run(&mut world);
+    /// # assert!(!message_changed.run((), &mut world));
+    /// ```
+    fn clone_if_neq<T>(&mut self, value: &T) -> bool
```
Comment (BenjaminBrienen) [inline crates/bevy_ecs/src/change_detection.rs:304]: I would prefer an enum with 2 variants which are self describing

**Review by BenjaminBrienen — APPROVED** [review]

File: crates/bevy_ecs/src/change_detection.rs (hunk 1 above)
Comment (SOF3) [reply crates/bevy_ecs/src/change_detection.rs:304]: You mean the return value? This is just to be consistent with `set_if_neq`. If we were to add an enum for that, I would prefer doing that in a separate PR and change `set_if_neq` together.

**Review by alice-i-cecile — APPROVED** [review]
I really appreciate the motivation / doc tests on this method: my immediate reaction was "how is this useful"!

**Comment by chescock** [comment]
Since this behaves more like `clone_from()` than `clone()`, would it make more sense to call it `clone_from_if_neq`?

**Comment by BenjaminBrienen** [comment]
CI is still failing

**Comment by alice-i-cecile** [comment]
Agreed, I'd prefer `clone_from_if_neq`. CI failure looks like a missed feature flag.

**Review by BenjaminBrienen — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@SOF3 `use alloc::borrow::ToOwned;` fixed the CI task for me locally. I would push to your branch to add it but I don't have permissions. Ping me when that's done and I'll get this merged.

**Comment by SOF3** [comment]
image
@alice-i-cecile edits are open to maintainers, not sure why you can't push. I will update it myself later today anyway.

**Comment by alice-i-cecile** [comment]
Hmm, very odd. Thanks!

**Comment by SOF3** [comment]
um the CI failure doesn't look relevant to this PR

**Comment by BenjaminBrienen** [comment]
> um the CI failure doesn't look relevant to this PR

#17060

**Pull request merged**